### ISSUE TITLE
[core] Gradle: check should fail for unsupported Java version

### ISSUE THREAD

**Issue by C-Otto** [pmd/pmd#2415]
**Affects PMD Version:** 

6.21.0 (and, most likely with Java 15, also 6.22.0)

**Description:**

If the Java version is unsupported, `gradle pmdMain` yields a successful build despite buggy code and `ignoreFailures = false`. I experienced this with PMD 6.21.0 and Java 14.

I only found the stacktrace after spending an hour trying to find the error in my configuration (configuring rulesets etc.). Instead of (silently) passing the build, I expect PMD to complain as loud as possible.

Note: I now know that 6.22.0 fixes the issue for Java 14. However, I suspect users will have the same problem once they try using Java 15.

**Exception Stacktrace:**
Only visible in `reports/pmd/main.html`:

```
net.sourceforge.pmd.PMDException: Error while processing X/SomeClass.java
Caused by: java.lang.IllegalArgumentException: Unsupported class file major version 58
```

**Steps to reproduce:**

1. Setup Gradle project with Java 14 and PMD 6.21.0 (or below)
2. Configure code/rules so that PMD should complain
3. Run `gradle pmdMain`

**Comment by jsotuyod** [comment]
@C-Otto I fear we can't help with this.

Gradle runs PMD through the Ant task. The PMD Ant task does support a property `failOnError` which when set to `true` will do exactly what you want: fail if any errors were encountered during processing. Unfortunately, Gradle is making no use of that property, nor checking the error count in any other way, limiting itself to have failures if a certain number of violations is found.

You should report this on Gradle's issue tracker. If they need support from us / would like to improve how the Ant task exposes this behavior it's definitely something we can discuss, but this is not something for us to change on our own.

**Comment by C-Otto** [comment]
I disagree. The logic surrounding `failOnError` is implemented in PMD, not in Gradle. It rather looks like the error counter is not increased in all error situations (like the one I described in the issue).

https://github.com/pmd/pmd/blob/a28943ae1b2e25d3aecb08108fdf1376e23d50f7/pmd-ant/src/main/java/net/sourceforge/pmd/ant/internal/PMDTaskImpl.java#L146-L148

https://github.com/pmd/pmd/blob/a28943ae1b2e25d3aecb08108fdf1376e23d50f7/pmd-core/src/main/java/net/sourceforge/pmd/util/log/internal/MessageReporterBase.java#L86-L88

**Comment by jsotuyod** [comment]
> I disagree. The logic surrounding `failOnError` is implemented in PMD, not in Gradle. It rather looks like the error counter is not increased in all error situations (like the one I described in the issue).

Yes, the counter is implemented by PMD, but as you pointed out in the code…

> https://github.com/pmd/pmd/blob/a28943ae1b2e25d3aecb08108fdf1376e23d50f7/pmd-ant/src/main/java/net/sourceforge/pmd/ant/internal/PMDTaskImpl.java#L146-L148

The counter could be up to 1 million, if `failOnError` is false, it will not throw [as per the Ant task documentation](https://docs.pmd-code.org/latest/pmd_userdocs_tools_ant.html#parameters):

> `failonerror`	Whether or not to fail the build if any errors occur while processing the files

And here you can see Gradle never sets the value of `failOnError`, so it remains at the default value of `false`

https://github.com/gradle/gradle/blob/7027ffa44b168afb77c8060cf772804de4201cec/platforms/jvm/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/PmdInvoker.groovy#L119-L149

Being false, Gradle will never get an exception running PMD due to processing errors.

It's up to Gradle to opt-in on wanting the task to throw if such errors occur, and fail the build. PMD fully supports it.

**Comment by C-Otto** [comment]
Thanks, I created a corresponding feature request.
